blaze-client/src/main/scala/org/http4s/blaze/client/PoolManager.scala

```diff
@@ -111,24 +114,30 @@ private final class PoolManager[F[_], A <: Connection[F]](
   private def incrConnection(key: RequestKey): F[Unit] =
     F.delay {
       curTotal.getAndIncrement()
-      allocated.update(key, allocated.getOrElse(key, 0) + 1)
+      val update: BiFunction[RequestKey, Int, Int] = (_: RequestKey, v: Int) => v + 1
+      allocated.compute(key, update)
+      ()
     }
 
   private def decrConnection(key: RequestKey): F[Unit] =
     F.delay {
       curTotal.getAndDecrement()
-      val numConnections = allocated.getOrElse(key, 0)
+      val numConnections = allocated.getOrDefault(key, 0)
       // If there are no more connections drop the key
       if (numConnections == 1) {
         allocated.remove(key)
         idleQueues.remove(key)
         ()
-      } else
-        allocated.update(key, numConnections - 1)
+      } else {
+        allocated.put(key, numConnections - 1)
+        ()
+      }
```

Comment:
If it is possible that this piece of code can be accessed from multiple threads simultaneously, then (I guess) it is not going to be thread-safe even if change each collection to a thread-safe version.

Consider the scenario:
1. Two threads (A and B) get to line 125 and both receive `numConnections == 2` for the same key X.
2. Consequently they both pass to line 132 directly and each of them decrease numConnection by 1 therefore it becomes 0 (skipping 1)
3. The key X never gets removed.

Wdyt?